Mathlib/Tactic/Find.lean

```diff
@@ -39,96 +101,272 @@ private partial def matchHyps : List Expr → List Expr → List Expr → MetaM
-lemmas which are `apply`able against the current goal.
+    forallTelescopeReducing cTy fun cParams cTy' ↦ do
+      if head == cTy'.toHeadIndex then
+        let pat := pat.expr.instantiateLevelParamsArray pat.paramNames
+          (← mkFreshLevelMVars pat.numMVars).toArray
+        let (_, _, pat) ← lambdaMetaTelescope pat
+        let (patParams, _, pat) ← forallMetaTelescopeReducing pat
+        isDefEq cTy' pat <&&> matchHyps patParams.toList [] cParams.toList
+      else
+        pure false
+
+end Mathlib.Tactic.Find
+
+
+/-!
+## A data structure for a relation on names
 -/
-elab "#find " t:term : command =>
-  liftTermElabM do
-    let t ← Term.elabTerm t none
-    Term.synthesizeSyntheticMVars (mayPostpone := false) (ignoreStuckTC := true)
-    findType t
-
-/- (Note that you'll get an error trying to run these here:
-   ``cannot evaluate `[init]` declaration 'findDeclsPerHead' in the same module``
-   but they will work fine in a new file!) -/
--- #find _ + _ = _ + _
--- #find _ + _ = _ + _
--- #find ?n + _ = _ + ?n
--- #find (_ : Nat) + _ = _ + _
--- #find Nat → Nat
--- #find ?n ≤ ?m → ?n + _ ≤ ?m + _
-
-open Lean.Elab.Tactic
-/-
-Display theorems (and definitions) whose result type matches the current goal,
-i.e. which should be `apply`able.
-```lean
-example : True := by find
-```
-`find` will not affect the goal by itself and should be removed from the finished proof.
-For a command that takes the type to search for as an argument,
-see `#find`, which is also available as a tactic.
+
+namespace Lean
+
+/-- A `NameRel` maps names to sets of names -/
+def NameRel := NameMap NameSet
+
+instance : EmptyCollection NameRel :=
+  inferInstanceAs $ EmptyCollection (NameMap NameSet)
+
+instance : Inhabited NameRel :=
+  inferInstanceAs $ Inhabited (NameMap NameSet)
+
+/-- For all names `n` mentioned in the type of the constant `c`, add `c.name` to the set associated
+with `n`. -/
+def NameRel.addDecl (c : ConstantInfo) (m : NameRel) := do
+  if ← c.name.isBlackListed then
+    return m
+  let consts := c.type.foldConsts {} (flip NameSet.insert)
+  return consts.fold (init := m) fun m n =>
+    m.insert n (
+      m.findD n {} |> flip .insert c.name
+    )
+
+end Lean
+
+
+/-!
+## The find tactic engine: Cache and matching
 -/
-elab "find" : tactic => do
-  findType (← getMainTarget)
 
-/-
+namespace Mathlib.Tactic.Find
+
+/-- The declaration cache used by `#find`, stores  `NameRel` mapping names to the
+name of constants they are mentinend in.
+
+The first `NameRel` is for library declaration (and doesn't change once populated),
+the second is for local declarations and is rebuilt upon every invocation of `#find`.  -/
+initialize findDeclsByConsts : DeclCache (NameRel × NameRel) ←
+  DeclCache.mk
+    (profilingName := "#find: init cache")
+    (empty := ({}, {}))
+    (addLibraryDecl := fun _ c (m₁, m₂) ↦ return (← NameRel.addDecl c m₁, m₂))
+    (addDecl := fun _ c (m₁, m₂) ↦ return (m₁, ← NameRel.addDecl c m₂))
```

Comment:
We really should have a global cache of this data, generated by CI, as for `exact?` / `rw?`.

This could happen in a follow-up PR.

Reply:
Agreed, and better in a follow up. I need to make sure the command isn't confused when the global cache refers to lemmas that are not locally available, given that the current environment may not have them yet.